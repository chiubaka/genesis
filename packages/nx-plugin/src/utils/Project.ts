import {
  getWorkspaceLayout,
  names,
  ProjectType as NxProjectType,
  Tree,
} from "@nrwl/devkit";
import path from "node:path";

export type ProjectType = NxProjectType | "e2e";

export interface ProjectNames {
  camelCase: string;
  kebabCase: string;
  pascalCase: string;
}

export class Project {
  private tree: Tree;
  private names: ProjectNames;
  private type: ProjectType;

  constructor(tree: Tree, name: string, type: ProjectType) {
    this.tree = tree;
    this.type = type;

    const {
      className: pascalCase,
      fileName: kebabCase,
      propertyName: camelCase,
    } = names(name);

    this.names = {
      camelCase,
      kebabCase,
      pascalCase,
    };
  }

  public path(relativePath = "") {
    return this.baseDir(path.join(this.getName(), relativePath));
  }

  public relativePath(relativePath = "") {
    const baseDirName = path.basename(this.baseDir());

    return path.join(baseDirName, this.getName(), relativePath);
  }

  public srcPath(relativePath = "") {
    return path.join(this.path("src"), relativePath);
  }

  public testPath(relativePath = "") {
    return path.join(this.path("test"), relativePath);
  }

  public distPath(relativePath = "") {
    return path.join("dist", this.relativePath(relativePath));
  }

  public jestConfigPath() {
    return this.path("jest.config.ts");
  }

  public getTree() {
    return this.tree;
  }

  public getScope() {
    const { npmScope } = getWorkspaceLayout(this.tree);
    return npmScope;
  }

  public getName() {
    return this.names.kebabCase;
  }

  public getNames() {
    return this.names;
  }

  public getType() {
    return this.type;
  }

  private baseDir(relativePath = "") {
    const { appsDir, libsDir } = getWorkspaceLayout(this.tree);

    const isPackagesLayout = path.basename(libsDir) === "packages";

    let baseDir = libsDir;
    if (
      (isPackagesLayout && this.type === "e2e") ||
      (!isPackagesLayout && this.type === "application")
    ) {
      baseDir = appsDir;
    }

    return path.join(baseDir, relativePath);
  }
}
