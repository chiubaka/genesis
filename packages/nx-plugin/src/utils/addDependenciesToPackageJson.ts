import {
  addDependenciesToPackageJson as nxAddDependenciesToPackageJson,
  Tree,
} from "@nrwl/devkit";

import { getLatestPackageVersion } from "./getLatestPackageVersion";

type Dependencies = string[] | DependenciesWithVersions;
type DependenciesWithVersions = Record<string, string | undefined>;

export const addDependenciesToPackageJson = async (
  tree: Tree,
  dependencies: Dependencies,
  devDependencies: Dependencies,
) => {
  const dependenciesWithVersions = await calculateDependencyVersions(
    dependencies,
  );
  const devDependenciesWithVersions = await calculateDependencyVersions(
    devDependencies,
  );

  return nxAddDependenciesToPackageJson(
    tree,
    dependenciesWithVersions,
    devDependenciesWithVersions,
  );
};

const calculateDependencyVersions = async (
  dependencies: Dependencies,
): Promise<Record<string, string>> => {
  const dependenciesWithVersions: Record<string, string> = {};

  if (isDependencyArray(dependencies)) {
    for (const dependencyName of dependencies) {
      const version = await getLatestPackageVersion(dependencyName);
      // eslint-disable-next-line security/detect-object-injection
      dependenciesWithVersions[dependencyName] = version;
    }
  } else {
    for (const dependencyName in dependencies) {
      const version =
        // eslint-disable-next-line security/detect-object-injection
        dependencies[dependencyName] ||
        (await getLatestPackageVersion(dependencyName));
      // eslint-disable-next-line security/detect-object-injection
      dependenciesWithVersions[dependencyName] = version;
    }
  }

  return dependenciesWithVersions;
};

const isDependencyArray = (
  dependencies: Dependencies,
): dependencies is string[] => {
  return Array.isArray(dependencies);
};
