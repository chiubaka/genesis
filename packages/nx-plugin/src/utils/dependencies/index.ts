export * from "./addDependenciesToPackageJson";
export * from "./compatiblePackageVersions";
