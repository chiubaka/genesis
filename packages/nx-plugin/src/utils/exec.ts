import { exec as originalExec } from "node:child_process";
import { promisify } from "node:util";

export const exec = promisify(originalExec);
