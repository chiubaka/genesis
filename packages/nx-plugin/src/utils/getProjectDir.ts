import { getWorkspaceLayout, ProjectType, Tree } from "@nrwl/devkit";
import path from "node:path";

export const getProjectDir = (
  tree: Tree,
  projectName: string,
  projectType: ProjectType,
) => {
  const { appsDir, libsDir } = getWorkspaceLayout(tree);
  const baseDir = projectType === "application" ? appsDir : libsDir;

  return path.join(baseDir, projectName);
};
