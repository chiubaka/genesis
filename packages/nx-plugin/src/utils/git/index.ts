export * from "./getGitHubRepoInfo";
export * from "./parseRemoteUrl";
