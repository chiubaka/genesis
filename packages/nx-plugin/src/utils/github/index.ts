import { GitHubApiAdapter } from "./GitHubApiAdapter";

export {
  BranchProtection as GitHubBranchProtection,
  Label as GitHubLabel,
  Repo as GitHubRepo,
} from "./GitHubApiAdapter";

export const github = new GitHubApiAdapter();
