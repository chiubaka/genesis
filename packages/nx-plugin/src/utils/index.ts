export * from "./addDependenciesToPackageJson";
export * from "./exec";
export * from "./getProjectDir";
export * from "./git";
export * from "./github";
export * from "./Project";
export * from "./projectNameCasings";
export * from "./tasks";
