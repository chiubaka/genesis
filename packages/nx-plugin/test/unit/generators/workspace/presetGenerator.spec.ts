import { NxJsonConfiguration, readJson, Tree } from "@nrwl/devkit";
import { createTreeWithEmptyWorkspace } from "@nrwl/devkit/testing";
import { PackageJson } from "nx/src/utils/package-json";

import {
  presetGenerator,
  PresetGeneratorSchema,
} from "../../../../src/generators";

describe("presetGenerator", () => {
  let tree: Tree;
  const options: PresetGeneratorSchema = {
    workspaceName: "preset",
    workspaceScope: "chiubaka",
    description: "Testing for the preset generator",

    disableImmutableInstalls: false,
    skipInstall: true,
  };

  beforeAll(async () => {
    tree = createTreeWithEmptyWorkspace();
    tree.write("apps/.gitkeep", "");
    tree.write("libs/.gitkeep", "");
    await presetGenerator(tree, options);
  });

  describe("package.json", () => {
    let packageJson: PackageJson;

    beforeAll(() => {
      packageJson = readJson<PackageJson>(tree, "package.json");
    });

    it("sets the root package name to the workspace name", () => {
      expect(packageJson["name"]).toBe("preset");
    });

    it("sets up yarn workspaces to pick up packages in the packages dir", () => {
      expect(packageJson.workspaces).toEqual({
        packages: ["packages/*"],
      });
    });

    it("matches snapshot", () => {
      expect(packageJson).toMatchSnapshot();
    });

    describe("scripts", () => {
      let scripts: Record<string, string> | undefined;

      beforeAll(() => {
        scripts = packageJson.scripts;
      });

      it("includes a build:affected script", () => {
        expect(scripts?.["build:affected"]).toBe("nx affected --target=build");
      });

      it("includes a build:ci script", () => {
        expect(scripts?.["build:ci"]).toBe(
          "yarn build:affected --base=$NX_BASE --head=$NX_HEAD",
        );
      });

      it("includes a test:all script", () => {
        expect(scripts?.["test:all"]).toBe("nx run-many --target=test --all");
      });

      it("includes a test:affected script", () => {
        expect(scripts?.["test:affected"]).toBe("nx affected --target=test");
      });

      it("includes a test:ci script", () => {
        expect(scripts?.["test:ci"]).toBe("./scripts/test-ci.sh");
      });

      it("includes a test:ci:affected script", () => {
        expect(scripts?.["test:ci:affected"]).toBe(
          "yarn test:affected --ci --coverage --base=$NX_BASE --head=$NX_HEAD && yarn e2e:affected --ci --coverage --base=$NX_BASE --head=$NX_HEAD",
        );
      });

      it("includes a test:ci:all script", () => {
        expect(scripts?.["test:ci:all"]).toBe(
          "yarn test:all --ci --coverage && yarn e2e:all --ci --coverage",
        );
      });

      it("includes an e2e script", () => {
        expect(scripts?.["e2e"]).toBe("nx e2e");
      });

      it("includes a e2e:all script", () => {
        expect(scripts?.["e2e:all"]).toBe("nx run-many --target=e2e --all");
      });

      it("includes a e2e:affected script", () => {
        expect(scripts?.["e2e:affected"]).toBe("nx affected --target=e2e");
      });

      it("includes a deploy script", () => {
        expect(scripts?.["deploy"]).toBe(
          "nx deploy $@ --configuration=production",
        );
      });

      it("includes a deploy:ci script", () => {
        expect(scripts?.["deploy:ci"]).toBe("yarn deploy");
      });

      it("creates a test-ci.sh file in the scripts directory", () => {
        expect(tree.exists("scripts/test-ci.sh")).toBe(true);
      });
    });
  });

  describe("workspace layout", () => {
    it("should create an e2e dir", () => {
      expect(tree.exists("e2e")).toBe(true);
    });

    it("should create a packages dir", () => {
      expect(tree.exists("packages")).toBe(true);
    });

    it("should not create an apps dir", () => {
      expect(tree.exists("apps")).toBe(false);
    });

    it("should not create a libs dir", () => {
      expect(tree.exists("libs")).toBe(false);
    });

    it("should update workspaceLayout in nx.json", () => {
      const nxJson = readJson<NxJsonConfiguration>(tree, "nx.json");

      expect(nxJson.workspaceLayout).toEqual({
        appsDir: "e2e",
        libsDir: "packages",
      });
    });
  });

  describe("testing", () => {
    it("generates a Codecov configuration file", () => {
      expect(tree.exists("codecov.yml")).toBe(true);
    });
  });

  describe("CI", () => {
    it("generates a .circleci/config.yml file", () => {
      expect(tree.exists(".circleci/config.yml")).toBe(true);
    });
  });
});
