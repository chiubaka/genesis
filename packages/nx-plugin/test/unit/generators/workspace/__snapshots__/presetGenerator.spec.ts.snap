// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`presetGenerator package.json matches snapshot 1`] = `
Object {
  "bugs": Object {
    "url": "https://github.com/chiubaka/preset/issues",
  },
  "dependencies": Object {},
  "devDependencies": Object {
    "@chiubaka/eslint-config": "^0.6.3",
    "@chiubaka/tsconfig": "^mocked",
    "@nrwl/eslint-plugin-nx": "15.4.8",
    "@nrwl/jest": "15.4.8",
    "eslint": "^mocked",
    "husky": "^mocked",
    "jest": "^28.1.3",
    "lint-staged": "^mocked",
    "prettier": "^2.8.8",
    "ts-jest": "^28.0.8",
    "tslib": "^mocked",
  },
  "homepage": "https://github.com/chiubaka/preset#readme",
  "name": "preset",
  "repository": Object {
    "type": "git",
    "url": "git+ssh://git@github.com/chiubaka/preset.git",
  },
  "scripts": Object {
    "build:affected": "nx affected --target=build",
    "build:ci": "yarn build:affected --base=$NX_BASE --head=$NX_HEAD",
    "deploy": "nx deploy $@ --configuration=production",
    "deploy:ci": "yarn deploy",
    "e2e": "nx e2e",
    "e2e:affected": "nx affected --target=e2e",
    "e2e:all": "nx run-many --target=e2e --all",
    "lint": "nx lint",
    "lint:affected": "nx affected --target=lint",
    "lint:all": "yarn lint:root && yarn lint:packages",
    "lint:ci": "yarn lint:root && yarn lint:affected --base=$NX_BASE --head=$NX_HEAD",
    "lint:fix:all": "yarn lint:fix:root; yarn lint:fix:packages",
    "lint:fix:packages": "yarn lint:packages --fix",
    "lint:fix:root": "yarn lint:root --fix",
    "lint:packages": "nx run-many --target=lint --all",
    "lint:root": "yarn exec eslint .",
    "lint:staged": "lint-staged",
    "prepare": "husky install",
    "test:affected": "nx affected --target=test",
    "test:all": "nx run-many --target=test --all",
    "test:ci": "./scripts/test-ci.sh",
    "test:ci:affected": "yarn test:affected --ci --coverage --base=$NX_BASE --head=$NX_HEAD && yarn e2e:affected --ci --coverage --base=$NX_BASE --head=$NX_HEAD",
    "test:ci:all": "yarn test:all --ci --coverage && yarn e2e:all --ci --coverage",
  },
  "workspaces": Object {
    "packages": Array [
      "packages/*",
    ],
  },
}
`;
